src/frontend/src/optimizer/rule/apply_hop_window_transpose_rule.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_pb::plan_common::JoinType;
+
+use super::{BoxedRule, Rule};
+use crate::optimizer::plan_node::{LogicalApply, LogicalFilter, LogicalHopWindow};
+use crate::optimizer::PlanRef;
+use crate::utils::Condition;
+
+/// Transpose `LogicalApply` and `LogicalHopWindow`.
+///
+/// Before:
+///
+/// ```text
+///     LogicalApply
+///    /            \
+///  Domain      LogicalHopWindow
+///                  |
+///                Input
+/// ```
+///
+/// After:
+///
+/// ```text
+///    LogicalHopWindow
+///          |
+///    LogicalApply
+///    /            \
+///  Domain        Input
+/// ```
+pub struct ApplyHopWindowTransposeRule {}
+impl Rule for ApplyHopWindowTransposeRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply: &LogicalApply = plan.as_logical_apply()?;
+        let (left, right, on, join_type, correlated_id, correlated_indices, max_one_row) =
+            apply.clone().decompose();
+        let hop_window: &LogicalHopWindow = right.as_logical_hop_window()?;
+        assert_eq!(join_type, JoinType::Inner);
+
+        if !hop_window.output_indices_is_trivial() {
```

Comment:
In which case the output can be non-trivial? I am wondering whether it's a very corner case, so that we can use this `if so return none` way.

Reply:
`output_indices` somehow can be regarded as a projection inside an operator. Trivial here means we have an identical project. We only handle hop window with trivial `output_indices` in this rule. I think we should add a rule to extract that project from the hop window before applying the subquery unnesting rules (just like how we treat `output_indices` inside join ).

Reply:
@fuyufjh could you PTAL again?

Reply:
Sure.

In my previous comment, I was wondering which case can lead to a Hop with non-trivial output_indices